# Load data

In [ ]:
import numpy as np
import pandas as pd
def load_data(file_path):
  data = np.load("./P12data/split_1/" + file_path, allow_pickle=True)

  # Let's create lists to hold the extracted data
  ts_values_list = []
  ts_indicators_list = []
  ts_times_list = []
  static_features_list = []
  labels_list = []

  # Iterate through all records to collect the data
  for record in data:
      # Extract the timeseries values, indicators, and timestamps
      ts_values = record['ts_values']
      ts_indicators = record['ts_indicators']
      ts_times = record['ts_times']

      # Flatten the ts_values if it's a 2D array, or leave it as is if 1D
      #ts_values_flat = ts_values.flatten() if ts_values.ndim > 1 else ts_values

      # Extract the static features and labels (assuming they're 1D arrays)
      static_features = record['static']
      labels = record['labels']

      # Append the extracted values to the respective lists
      ts_values_list.append(ts_values)
      ts_indicators_list.append(ts_indicators)
      ts_times_list.append(ts_times)
      static_features_list.append(static_features)
      labels_list.append(labels)

  # Create a DataFrame from the collected lists
  df = pd.DataFrame({
      'ts_values': ts_values_list,
      'ts_indicators': ts_indicators_list,
      'ts_times': ts_times_list,
      'static': static_features_list,
      'labels': labels_list
  })
  df['ts_indicators'] = df['ts_indicators'].apply(lambda x: np.array(x, dtype=int))
  df['ts_indicators']
  return df

In [2]:
# Load the data from the .npy file
file_path = '../P12data/split_1/train_physionet2012_1.npy'
df_train = load_data(file_path)
# Display the first few rows of the DataFrame
df_train.head()

FileNotFoundError: [Errno 2] No such file or directory: 'P12data/split_1/train_physionet2012_1.npy'

In [7]:
# Load the data from the .npy file
file_path = '../P12data/split_1/test_physionet2012_1.npy'
df_test = load_data(file_path)
# Display the first few rows of the DataFrame
df_test.head()

ts_values  \
0  [[0.16966099858624017, 0.0, 0.0, 0.0, 0.0, 0.0...   
1  [[-0.32848355450987454, 0.0, 0.0, 0.0, 0.0, 0....   
2  [[-0.9359767612187836, 0.0, 0.0, 0.0, 0.0, 0.0...   
3  [[-0.9481265944542595, 0.0, 0.0, 0.0, 0.0, 0.0...   
4  [[-0.5755307301074263, 0.0, 0.0, 0.0, 0.0, 0.0...   

                                       ts_indicators  \
0  [[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   
1  [[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   
2  [[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   
3  [[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   
4  [[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   

                                            ts_times  \
0  [0.0, 1.0166667, 1.9166666, 2.95, 3.7, 3.78333...   
1  [0.0, 0.15, 0.65, 1.65, 2.65, 3.15, 3.4, 4.15,...   
2  [0.0, 0.31666666, 0.85, 1.25, 2.1, 2.35, 2.6, ...   
3  [0.0, 0.15, 0.23333333, 0.4, 0.65, 1.15, 1.65,...   
4  [0.0, 0.1, 1.1, 2.1, 2.15, 3.1, 4.1, 5.1, 6.1,...   

                                              static  labels  
0  [0.3201678628319356, 0.0, 1.0, 1.0059237867031...       0  
1  [0.5540780043938743, 0.0, 1.0, 1.0059237867031...       0  
2  [-0.2646074910729112, 1.0, 0.0, -1.04133971147...       0  
3  [-1.2587255927111507, 0.0, 1.0, -1.04133971147...       0  
4  [-0.03069734951097247, 1.0, 0.0, 0.97573151501...       0

In [8]:
# Load the data from the .npy file
file_path = '../P12data/split_1/validation_physionet2012_1.npy'
df_val = load_data(file_path)
# Display the first few rows of the DataFrame
df_val.head()

ts_values  \
0  [[-0.6079304913933774, 0.0, 0.0, 0.0, 0.0, 0.0...   
1  [[-1.2599731381302017, 0.0, 0.0, 0.0, 0.0, 0.0...   
2  [[1.2104992358193618, 0.0, 0.0, 0.0, 0.0, 0.0,...   
3  [[-3.406449135168347, 0.0, 0.0, 0.0, 0.0, 0.0,...   
4  [[-0.6686798120642683, 0.0, 0.0, 0.0, 0.0, 0.0...   

                                       ts_indicators  \
0  [[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   
1  [[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   
2  [[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   
3  [[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   
4  [[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   

                                            ts_times  \
0  [0.0, 0.11666667, 0.85, 0.93333334, 1.0166667,...   
1  [0.0, 0.4, 0.65, 0.9, 1.4, 1.7833333, 1.9, 2.3...   
2  [0.0, 0.2, 0.26666668, 0.51666665, 0.76666665,...   
3  [0.0, 0.05, 0.3, 0.55, 1.55, 2.55, 3.55, 4.55,...   
4  [0.0, 0.05, 0.55, 1.05, 1.0666667, 1.3, 2.05, ...   

                                              static  labels  
0  [-0.2061299556824265, 0.0, 1.0, 1.005923786703...       0  
1  [-0.6154727034158193, 1.0, 0.0, 0.887477495879...       0  
2  [0.43712293361290494, 1.0, 0.0, 0.740000065545...       0  
3  [-0.9663379157587273, 1.0, 0.0, -1.04133971147...       0  
4  [1.2558084290796905, 0.0, 1.0, 0.9467005526959...       0

# Preprocess data

In [59]:
from sklearn.utils import resample

# Separate majority and minority classes
df_majority = df_train[df_train.labels == 0]
df_minority = df_train[df_train.labels == 1]

# Upsample minority class
df_minority_upsampled = resample(df_minority, 
                 replace=True,     # sample with replacement
                 n_samples=int(len(df_majority)*40/60),    # to match majority class
                 random_state=123) # reproducible results

# Combine majority class with upsampled minority class
df_train_upsampled = pd.concat([df_majority, df_minority_upsampled])

# Display new class counts
print(df_train_upsampled.labels.value_counts()/len(df_train_upsampled))

labels
0    0.600029
1    0.399971
Name: count, dtype: float64


In [56]:
df_train = df_train_upsampled

# Rest

In [67]:
df_train_saved = df_train

In [75]:
df_train = df_train_saved.sample(frac=0.05).reset_index(drop=True)
len(df_train)

687

In [49]:
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, Dataset, random_split
from sklearn.metrics import accuracy_score
from torch.nn.utils.rnn import pad_sequence, pack_padded_sequence, pad_packed_sequence
import numpy as np

In [94]:
# Parameters based on your data
max_sequence_length_train = df_train['ts_times'].apply(len).max()  # Max sequence length based on ts_times
max_sequence_length_test = df_test['ts_times'].apply(len).max()  # Max sequence length based on ts_times
max_sequence_length_val = df_val['ts_times'].apply(len).max()

num_sensors = 2                    # Number of sensors (now 2)
num_measurements_per_sensor = 37   # Measurements per sensor per timestamp
num_static_features = 8            # Static features (age, height, etc.)
num_classes = 2                    # Number of classes (0 or 1)

# New temporal input size (2 sensors with 37 measurements each)
input_size = num_sensors * num_measurements_per_sensor  # 2 * 37 = 74

# Hyperparameters
hidden_size = 64
num_layers = 1
num_epochs = 100
batch_size = 8
learning_rate = 0.001

In [100]:
# Adjusted dataset class for your data
class PatientDataset(Dataset):
    def __init__(self, df, max_sequence_length, input_size, num_static_features, num_classes):
        self.df = df
        self.max_sequence_length = max_sequence_length
        self.input_size = input_size
        self.num_static_features = num_static_features
        self.num_classes = num_classes

        self.temporal_data = []
        self.static_data = []
        self.targets = []
        self.sequence_lengths = []

        # Extract data
        for idx, row in df.iterrows():
            ts_values = torch.tensor(row['ts_values'], dtype=torch.float32, device=torch.cuda.current_device())  # Temporal data (sensor 1)
            ts_indicators = torch.tensor(row['ts_indicators'], dtype=torch.float32, device=torch.cuda.current_device())  # Temporal data (sensor 2)
            ts_times = torch.tensor(row['ts_times'], dtype=torch.float32, device=torch.cuda.current_device())  # Time spans (timestamps)
            static_features = torch.tensor(row['static'], dtype=torch.float32, device=torch.cuda.current_device())  # Static features
            label = row['labels']  # Class label

            seq_len = len(ts_times)  # Sequence length
            self.sequence_lengths.append(seq_len)

            # Concatenate ts_values and ts_indicators to create a combined input for both sensors
            combined_temporal_data = torch.cat((ts_values, ts_indicators), dim=-1)  # shape: [seq_len, 74]
            self.temporal_data.append(combined_temporal_data)
            self.static_data.append(static_features)
            self.targets.append(label)

    def __len__(self):
        return len(self.df)

    def __getitem__(self, idx):
        return self.temporal_data[idx], self.static_data[idx], self.targets[idx], self.sequence_lengths[idx]

# Split dataset into train, validation, and test sets
train_data = PatientDataset(df_train, max_sequence_length_train, input_size, num_static_features, num_classes)
test_data = train_data
val_data = train_data

# Custom collate function to pad sequences
def collate_fn(batch):
    temporal_data, static_data, targets, seq_lengths = zip(*batch)
    seq_lengths = torch.tensor(seq_lengths, device=torch.cuda.current_device())
    static_data = torch.stack(static_data)
    targets = torch.tensor(targets, device=torch.cuda.current_device())

    # Pad temporal data sequences to match the longest sequence in the batch
    temporal_data_padded = pad_sequence(temporal_data, batch_first=True)
    return temporal_data_padded, static_data, targets, seq_lengths

# Create DataLoader instances
train_loader = DataLoader(train_data, batch_size=batch_size, shuffle=True, collate_fn=collate_fn)
val_loader = DataLoader(val_data, batch_size=batch_size, shuffle=False, collate_fn=collate_fn)
test_loader = DataLoader(test_data, batch_size=batch_size, shuffle=False, collate_fn=collate_fn)



# Define a Deep State Space Model (DSSM) for classification
class DeepStateSpaceModel(nn.Module):
    def __init__(self, input_size, hidden_size, static_input_size, num_classes, num_layers=1):
        super(DeepStateSpaceModel, self).__init__()

        # LSTM for temporal data (this is the "observation model")
        self.lstm = nn.LSTM(input_size, hidden_size, num_layers, batch_first=True, device=torch.cuda.current_device())
        
        # Linear layer for static data (this is the "static feature processing")
        self.static_fc = nn.Linear(static_input_size, hidden_size, device=torch.cuda.current_device())

        # Transition matrix for the state space model (latent state evolution)
        self.transition_matrix = nn.Parameter(torch.randn(hidden_size, hidden_size) * 0.1)  # Scaled initialization

        # Fully connected layer for combined features
        self.combined_fc = nn.Linear(hidden_size * 2, hidden_size)

        # Output layer (for classification)
        self.output_layer = nn.Linear(hidden_size, num_classes, device=torch.cuda.current_device())  # Changed from hidden_size to hidden_size * 2

        self.first = True
        # Normalization layers
        self.norm_temporal = nn.LayerNorm(hidden_size)
        self.norm_static = nn.LayerNorm(hidden_size)
        self.norm_combined = nn.LayerNorm(hidden_size)

    def forward(self, temporal_data, static_data, seq_lengths):
        # Pack and pass temporal data through LSTM
        # temporal_data = self.norm_1(temporal_data)
        packed_input = pack_padded_sequence(temporal_data, seq_lengths.cpu(), batch_first=True, enforce_sorted=False)
        packed_output, (hn, _) = self.lstm(packed_input)

        # Initial latent state (from LSTM's final hidden state)
        latent_state = hn[-1, :, :]  # shape: [batch_size, hidden_size]

        # Apply non-linear transformation and normalization to latent state
        latent_state =  nn.functional.relu(self.norm_temporal(latent_state))

        # State space evolution: evolve the latent state over time using the transition matrix
        for t in range(1, temporal_data.size(1)):  # Loop through time steps (sequence length)
            latent_state = torch.matmul(latent_state, self.transition_matrix)  # Transition from previous state to the next

        # Process static data with a fully connected layer
        static_out = nn.functional.relu(self.norm_static(self.static_fc(static_data)))

        # Combine the LSTM output with static features
        combined = torch.cat((latent_state, static_out), dim=1)  # shape: [batch_size, hidden_size * 2]
        combined = nn.functional.relu(self.norm_combined(self.combined_fc(combined)))
        # Output classification
        # combined = self.norm_3(combined)
        output = self.output_layer(combined)  # shape: [batch_size, num_classes]

        return output

# Class weights for handling class imbalance
class_weights = torch.tensor([1.0, 5.0], device=torch.cuda.current_device())  # Adjust the weights based on your class imbalance
criterion = nn.CrossEntropyLoss(weight=class_weights)  # For classification with class imbalance

# Initialize model, loss function, and optimizer
model = DeepStateSpaceModel(input_size=input_size, hidden_size=hidden_size,
                            static_input_size=num_static_features, num_classes=num_classes, num_layers=num_layers)
model = model.to(torch.cuda.current_device())
optimizer = optim.Adam(model.parameters(), lr=learning_rate)

# Gradient clipping to avoid exploding gradients
max_grad_norm = 5.0  # Clip gradients to a maximum norm of 1.0

# Training loop with validation
for epoch in range(num_epochs):
    model.train()  # Set the model to training mode
    for batch in train_loader:
        temporal_batch, static_batch, target_batch, seq_lengths = batch

        # Zero the gradients
        optimizer.zero_grad()

        # Forward pass
        output = model(temporal_batch, static_batch, seq_lengths)

        loss = criterion(output, target_batch)  # CrossEntropyLoss expects raw logits

        # Backward pass and optimization
        loss.backward()

        # Clip gradients to prevent exploding gradients
        torch.nn.utils.clip_grad_norm_(model.parameters(), max_grad_norm)

        optimizer.step()

    # Validation loop to calculate accuracy on validation data
    model.eval()  # Set the model to evaluation mode
    val_preds = []
    val_labels = []
    with torch.no_grad():
        for batch in val_loader:
            temporal_batch, static_batch, target_batch, seq_lengths = batch
            output = model(temporal_batch, static_batch, seq_lengths)

            # Get predictions
            _, preds = torch.max(output, 1)
            val_preds.extend(preds.cpu().numpy())
            val_labels.extend(target_batch.cpu().numpy())

    # Calculate validation accuracy
    val_accuracy = accuracy_score(val_labels, val_preds)
    print(f"Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}, Validation Accuracy: {val_accuracy * 100:.2f}%")

# Final test accuracy calculation
model.eval()  # Set the model to evaluation mode
test_preds = []
test_labels = []
with torch.no_grad():
    for batch in test_loader:
        temporal_batch, static_batch, target_batch, seq_lengths = batch
        output = model(temporal_batch, static_batch, seq_lengths)

        # Get predictions
        _, preds = torch.max(output, 1)
        test_preds.extend(preds.cpu().numpy())
        test_labels.extend(target_batch.cpu().numpy())

# Calculate test accuracy
test_accuracy = accuracy_score(test_labels, test_preds)
print(f"Test Accuracy: {test_accuracy * 100:.2f}%")


Epoch [1/100], Loss: 0.2484, Validation Accuracy: 43.81%
Epoch [2/100], Loss: 0.8381, Validation Accuracy: 47.74%
Epoch [3/100], Loss: 0.4558, Validation Accuracy: 49.05%
Epoch [4/100], Loss: 0.6637, Validation Accuracy: 52.84%
Epoch [5/100], Loss: 0.6905, Validation Accuracy: 53.42%
Epoch [6/100], Loss: 0.5394, Validation Accuracy: 46.00%
Epoch [7/100], Loss: 0.3549, Validation Accuracy: 49.78%
Epoch [8/100], Loss: 0.2635, Validation Accuracy: 49.64%
Epoch [9/100], Loss: 0.4168, Validation Accuracy: 46.58%
Epoch [10/100], Loss: 0.4445, Validation Accuracy: 52.69%
Epoch [11/100], Loss: 0.4553, Validation Accuracy: 48.91%
Epoch [12/100], Loss: 0.5088, Validation Accuracy: 55.02%
Epoch [13/100], Loss: 0.4359, Validation Accuracy: 51.53%
Epoch [14/100], Loss: 0.5852, Validation Accuracy: 55.31%
Epoch [15/100], Loss: 0.5514, Validation Accuracy: 56.48%
Epoch [16/100], Loss: 0.4658, Validation Accuracy: 47.74%
Epoch [17/100], Loss: 0.5726, Validation Accuracy: 49.20%
Epoch [18/100], Loss: 0

- No batchnorm:
Epoch [1/10], Loss: 0.4598, Validation Accuracy: 40.32%
Epoch [2/10], Loss: 0.9196, Validation Accuracy: 43.38%
Epoch [3/10], Loss: 0.7145, Validation Accuracy: 43.23%
Epoch [4/10], Loss: 0.7549, Validation Accuracy: 46.58%
Epoch [5/10], Loss: 0.6851, Validation Accuracy: 47.31%
Epoch [6/10], Loss: 0.6319, Validation Accuracy: 46.72%
Epoch [7/10], Loss: 0.4454, Validation Accuracy: 49.78%
Epoch [8/10], Loss: 0.4319, Validation Accuracy: 50.07%
Epoch [9/10], Loss: 0.3745, Validation Accuracy: 49.78%
Epoch [10/10], Loss: 0.4979, Validation Accuracy: 53.42%
Test Accuracy: 53.42%
- Batchnorm 1:
Epoch [1/10], Loss: 0.8834, Validation Accuracy: 40.47%
Epoch [2/10], Loss: 0.4536, Validation Accuracy: 44.25%
Epoch [3/10], Loss: 0.5721, Validation Accuracy: 46.43%
Epoch [4/10], Loss: 0.4230, Validation Accuracy: 48.76%
Epoch [5/10], Loss: 0.8273, Validation Accuracy: 49.20%
Epoch [6/10], Loss: 0.3996, Validation Accuracy: 50.07%
Epoch [7/10], Loss: 0.4100, Validation Accuracy: 49.78%
Epoch [8/10], Loss: 0.4178, Validation Accuracy: 50.66%
Epoch [9/10], Loss: 0.6027, Validation Accuracy: 51.97%
Epoch [10/10], Loss: 0.3193, Validation Accuracy: 50.51%
Test Accuracy: 50.51%
- Batchnorm 2:
Epoch [1/10], Loss: 0.5966, Validation Accuracy: 51.82%
Epoch [2/10], Loss: 0.3539, Validation Accuracy: 51.09%
Epoch [3/10], Loss: 0.2553, Validation Accuracy: 48.76%
Epoch [4/10], Loss: 0.3089, Validation Accuracy: 49.49%
Epoch [5/10], Loss: 0.3994, Validation Accuracy: 51.67%
Epoch [6/10], Loss: 0.3683, Validation Accuracy: 49.93%
Epoch [7/10], Loss: 0.4936, Validation Accuracy: 50.07%
Epoch [8/10], Loss: 0.5954, Validation Accuracy: 50.51%
Epoch [9/10], Loss: 0.5542, Validation Accuracy: 50.36%
Epoch [10/10], Loss: 0.5420, Validation Accuracy: 50.22%
Test Accuracy: 50.22%
- Batchnorm 3:
Epoch [1/10], Loss: 0.6734, Validation Accuracy: 51.97%
Epoch [2/10], Loss: 0.6943, Validation Accuracy: 52.11%
Epoch [3/10], Loss: 1.3486, Validation Accuracy: 50.51%
Epoch [4/10], Loss: 0.4996, Validation Accuracy: 51.67%
Epoch [5/10], Loss: 0.3000, Validation Accuracy: 48.18%
Epoch [6/10], Loss: 0.5406, Validation Accuracy: 49.49%
Epoch [7/10], Loss: 0.4286, Validation Accuracy: 50.36%
Epoch [8/10], Loss: 0.8129, Validation Accuracy: 50.95%
Epoch [9/10], Loss: 0.7775, Validation Accuracy: 48.91%
Epoch [10/10], Loss: 0.5658, Validation Accuracy: 49.49%
Test Accuracy: 49.49%


In [29]:
df_train.labels.value_counts()

labels
0    8248
1    1342
Name: count, dtype: int64

In [26]:
print(val_preds)

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 

1. Padding: different time stamps length, to account for that we do padding with mask so the model knows to take only relevant measurements
2. Class imbalance: Given that the dataset has an imbalance of 86%, meaning one class represents 86% of the samples and the other class represents 14%, the class weights should be adjusted to compensate for the imbalance, such that the model gives more importance to the minority class (14%) and less importance to the majority class (86%). (try with values ([1.163, 7.143])
3. DSSM combining LSTM for input and SSM for output
4. Other hyperparameters to tune:
- hidden_size
- num_layers
- num_static_features
- transition_matrix
- batch_size
- learning_rate
- num_epochs
- max_grad_norm
- optimizer
- class_weights
5. We are predicting mortality so actually different metrics would be better to use



# Results step by step

1. Rebalance to 60/40. `Epoch [10/10], Loss: 1.0779, Validation Accuracy: 47.74% Test Accuracy: 47.74%`
2. 